Add support for relocations

janvrany opened the pull request:
This PR adds basic support for relocations. 

At the moment, this PR is work in progress. I'm still not 100% happy with it - design wise - but this should be enough to enable further experiments with TR and AoT withing ST25.

One thing that is probably missing is to add proper support for code object sections, so compiler can for example produce jump table for PICs in "data" section or big constants or global variable addresses  in "rodata". 

Once finalized, this would probably need to be split into 2 or three smallter PRs.

janvrany committed:
DSL: add support for relocations

This commit adds relocation support to DSL assembler. To to relocate
an operand, use `relocation_type % operand`, for example:

    asm auipc: ra, ra, R_RISCV_PCREL_HI20 % label

janvrany committed:
Add pool `AcIntLimits` containing min/max values of fixed-size ints.

This commit adds pool with `*_MIN` and `*_MAX` values for fixed-length
integers (both signed and unsigned). Using them rather than integer
literals improve readability.

janvrany committed:
Add pool `AcRISCVISALimits`

This commit adds pool containing various constants related to RISC-V
ISAm such as bit-size of immediates, min / max values of those and so
on to improve readability of code.

janvrany committed:
RISC-V: add `R_RISCV_CALL_PLT` relocation

See [1] for details.

[1]: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#relocations

janvrany committed:
DSL: add `AcDSLCodeBuffer >> size` to return size of the code in bytes

shingarov commented:
Hmm… looking at this code, I am now inclined to re-read Stephen Kell's [The Missing Link](https://dx.doi.org/10.1145/2983990.2983996), much more carefully this time.

janvrany committed:
DSL: add set of exported symbols to code object

This commit add `AcDSLCodeObject >> exports` that return symbols
"exported" by this code object, that is, a dictionary mapping
exported symbol to its address (respecting object's base address).
Each exported symbol must be paired with corresponding label in
instructions sequence.

To export a symbol, `AcDSLAssembler` provides method `global: symbol`
just like GNU assembler has .global directive [1].

[1]: https://sourceware.org/binutils/docs/as/Global.html

janvrany committed:
RISC-V: add `R_RISCV_PCREL_HI20` relocation

janvrany committed:
RISC-V: add `R_RISCV_PCREL_LO12_*` relocations

janvrany committed:
RISC-V: add `R_RISCV_32` and `R_RISCV_64` relocations

janvrany committed:
Add support for GOT

This commit add basic support for GOT accesses. During relocation
processing (linking) a relocation may request GOT entry for its symbol
by

    gotentry := AcRelocationRequestGOTEntry query: self

The response (`gotentry` in the above example) is an association from
GOT entry symbol to its address (that is, address of the entry). Actual
allocation of GOT entries is the responsibility of the linker.

janvrany committed:
RISC-V: add `R_RISCV_GOT_HI20` relocation

This commit adds `R_RISCV_GOT_HI20` relocation class and fixes
`R_RISCV_PCREL_LO12_*` to also work with it. This allows consumers to
generate GOT access for external symbols.

janvrany committed:
Add basic infrastructure for relocations

This commit adds basic support for relocation entries.

Each processor instruction can have a relocation entry. A relocation
entry is a subclass of `AcRelocation` (one subclass per relocation type)
and contains a (relocatable) symbol.

Particular relocation type (a concrete subclass of `AcRelocation`) is
valid only for some instructions (`#isValidForInstruction`) and knows
how to fixup (patch) that instruction, knowing the final address of its
symbol (note, that this final address might be address of trampoline or
into address table if final address is beyond the maximum displacement).

In some cases, relocation entries must come in pairs (like %lo must
follow %hi). Currently, there's no way to ensure nor validate that -
this is user's responsibility.

janvrany committed:
DSL: Add `AcDSLAssembler >> ptr:` to write pointer-size data

janvrany committed:
Add support for PLT

This commit add basic support for PLT calls, using similar mechanism
as for GOT entries introduced in previous commit.

janvrany committed:
DSL: add `#example02_riscv64_pic_code` demonstrating relocations

This commit add simple example showing how to use relocations when
using DSL assembler.
